### PR TITLE
hpx: Add conflict with some GCC versions and `+rocm` due to `valarray` bug

### PR DESCRIPTION
**Pull request by msimberg**
"Fixes" #38104 by adding conflicts with the buggy GCC versions and `+rocm`.

### CODE REVIEW

**Review by hkaiser — APPROVED** [review]
LGTM, thanks!

**Review by eugeneswalker — DISMISSED** [review]
Thank you!

**Comment by msimberg** [comment]
https://gitlab.spack.io/spack/spack/-/jobs/7346404 now fails to concretize because of the added conflicts. The GCC version seems to be fixed at 11.2.0. Is there a possibility to update that to 11.3.0?

**Comment by haampie** [comment]
Can you bump GCC @eugeneswalker? Maybe just use vanilla ubuntu 22.04?

**Comment by eugeneswalker** [comment]
> Can you bump GCC @eugeneswalker? Maybe just use vanilla ubuntu 22.04?

We can’t, no, because we have other issues that come up w newer gcc. But we can build hpx w gcc 11.1, so this conflict should not apply to gcc 11.1, which is what we use in the ci image. Removing that conflict will fix the CI failure.

**Comment by haampie** [comment]
What packages don't build with gcc 11.3 but do with 11.2? 🤔

**Comment by eugeneswalker** [comment]
> What packages don't build with gcc 11.3 but do with 11.2? 🤔

I have no idea. But since `hpx +rocm %gcc@11.1.0` builds fine -- we do not hit the valarray issue -- there is no real conflict. Merely removing the `conflict` from this PR will resolve the CI issue while preserving the true conflict with GCC 11.2.0. I can't justify the investment of time right now to actually do a new set of E4S builds (800-1000 packages per target) with GCC 11.3.0 on ARM, POWER, and X86_64 when it already works OK with GCC 11.1.0 which is easily provided by apt. I actually would like to do as you suggest and use a newer Spack-built GCC in future, just not right now.

**Review by eugeneswalker — APPROVED** [review]

**Comment by msimberg** [comment]
> > What packages don't build with gcc 11.3 but do with 11.2? thinking
> 
> I have no idea. But since `hpx +rocm %gcc@11.1.0` builds fine -- we do not hit the valarray issue -- there is no real conflict. Merely removing the `conflict` from this PR will resolve the CI issue while preserving the true conflict with GCC 11.2.0. I can't justify the investment of time right now to actually do a new set of E4S builds (800-1000 packages per target) with GCC 11.3.0 on ARM, POWER, and X86_64 when it already works OK with GCC 11.1.0 which is easily provided by apt. I actually would like to do as you suggest and use a newer Spack-built GCC in future, just not right now.

Thanks @eugeneswalker for checking that. I... assumed it would fail with 11.1.0 as well. If this passes CI I'm happy to go with this change. However, I'm slightly confused since a regular checkout of 11.1.0 definitely doesn't have the fix linked earlier. Are you sure the 11.1.0 you tested with doesn't happen to have been patched downstream?

**Comment by eugeneswalker** [comment]
> I'm slightly confused since a regular checkout of 11.1.0 definitely doesn't have the fix linked earlier. Are you sure the 11.1.0 you tested with doesn't happen to have been patched downstream?

Very good point. I don't know. It would not surprise me if the Ubuntu maintainers patched their GCC 11.1.0 before posting to apt repo.

**Comment by msimberg** [comment]
@eugeneswalker sorry for holding this up. I checked the GCC 11.1.0 that's in the e4s build image and it actually looks consistent wrt the error message in https://github.com/spack/spack/issues/38104, i.e. I can kind of see why it works with 11.1.0. The failure in https://github.com/spack/spack/issues/38104 with 11.2.0 is now more of a mystery to me (because _that_ one doesn't match what's in upstream libstdc++), but I'm unfortunately not going to look into this any deeper.

Let's go ahead with what's here so that the e4s stacks can start working. I suppose in most non-e4s cases users will have access to 11.3.0 or newer so this should anyway hopefully be a non-issue there. We can always refine the constraints if we get new information.

**Comment by eugeneswalker** [comment]
> Let's go ahead with what's here so that the e4s stacks can start working. I suppose in most non-e4s cases users will have access to 11.3.0 or newer so this should anyway hopefully be a non-issue there. We can always refine the constraints if we get new information.

That sounds great. Thank you!

**Comment by eugeneswalker** [comment]
@spackbot run pipeline

**Comment by scottwittenburg** [comment]
This needs fixes from a newer `develop`, but spackbot can't do that for you.  Can you rebase this on a `develop` on or after `[9eb3de85](https://gitlab.spack.io/spack/spack/-/commit/9eb3de85c5369260bc702ac5e1d257ff58aec4fe)`?  Doing that will allow the pipeline to run right away.  If you aren't in a hurry, you can also just push the merge with develop button.

**Pull request merged**
